### PR TITLE
Mask sensitive info in named collection's overrides

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category:
- Improvement

### Changelog entry:
Hide password in logs when a table function's arguments are specified with a named collection:

```
SELECT * FROM mysql(named_collection, host = 'mysql57', port = 3306, database = 'db', table = 'table',
                    user = 'user', password = 'password')
```

### CODE REVIEW

File: tests/integration/test_mask_sensitive_info/test.py [hunk 1]
```diff
@@ -116,6 +122,12 @@ def test_create_table():
         f"S3('http://minio1:9001/root/data/test3.csv.gz', 'CSV', 'gzip')",
         f"S3('http://minio1:9001/root/data/test4.csv', 'minio', '{password}', 'CSV')",
         f"S3('http://minio1:9001/root/data/test5.csv.gz', 'minio', '{password}', 'CSV', 'gzip')",
+        f"MySQL(named_collection_1, host = 'mysql57', port = 3306, database = 'mysql_db', table = 'mysql_table', user = 'mysql_user', password = '{password}')",
```
Comment (vitlibar) [inline tests/integration/test_mask_sensitive_info/test.py:125]: Normally in such cases the password is specified inside the name collection, however it's allowed to set it in a named argument like this. So we should recognize that case and hide the password.

File: src/Parsers/ASTFunction.cpp [hunk 2]
```diff
@@ -37,94 +37,130 @@ namespace
...
         }
 
-        std::pair<size_t, size_t> getRange() const
+        struct Result
         {
-            if (!arguments)
-                return npos;
+            /// Result constructed by default means no arguments will be hidden.
+            size_t start = static_cast<size_t>(-1);
+            size_t count = 0;
```
Comment (SmitaRKulkarni) [inline src/Parsers/ASTFunction.cpp:68]: For my understanding, what would be an example query where we have more than 1 field to be hidden ? 
Also I believe we assume that these field to be hidden will be consecutive arguments as we have a `start` & `count`, will it always be the case (excluding named collections) ?

Comment (SmitaRKulkarni) [reply src/Parsers/ASTFunction.cpp:68]: I see we use this for userid, password etc, and form the test examples looks like they are consecutive arguments.

Comment (SmitaRKulkarni) [reply src/Parsers/ASTFunction.cpp:68]: I see we have examples of `userid`, `password` etc and from the test examples looks like they are consecutive

File: src/Parsers/ASTFunction.cpp [hunk 3]
```diff
@@ -300,66 +347,123 @@ namespace
...
             {
                 /// BACKUP ... TO S3(url, [aws_access_key_id, aws_secret_access_key])
-                return {2, 3};
+                markSecretArgument(2);
             }
-            else
+        }
+
+        /// Whether a specified argument can be the name of a named collection?
+        bool isCollectionNameArgument(size_t arg_idx) const
```
Comment (SmitaRKulkarni) [inline src/Parsers/ASTFunction.cpp:432]: maybe you can call it `isNamedCollectionArgument` ? as in most places its referred to as `named_collection`. What do you think?

**Review by SmitaRKulkarni — APPROVED** [review]
Rest all LGTM

File: src/Parsers/ASTFunction.cpp (hunk 3 above)
Comment (vitlibar) [reply src/Parsers/ASTFunction.cpp:432]: In such cases all the arguments are related to the named collection:

```
func(named_collection_name, named_collection_override_param1 = val1, named_collection_override_param2 = val2, ...)
```

I think I'll rename the function to `isNamedCollectionNameArgument()` - it seems a bit long, but correct.

File: src/Parsers/ASTFunction.cpp (hunk 2 above)
Comment (vitlibar) [reply src/Parsers/ASTFunction.cpp:68]: Hidden arguments are always consecutive. Though in general they don't have to be always consecutive, they're consecutive in all the known cases.

There only few cases where we have more than 1 argument to be hidden:

```
encrypt('mode', 'plaintext', 'key' [, iv, aad]) -> encrypt('mode', '[HIDDEN]')
remote('addresses_expr', db_or_db_table, a1, a2, a3) -> remote('addresses_expr, db_table, a1, 'HIDDEN', 'HIDDEN')
mysql(named_collection_name, password = 'password', user = 'user', password = 'password')
```

The last two cases need explanation: for `remote()` we hide both `a2` and `a3` if we cannot easily determine which of `a2` and `a3` is a password. And for `mysql` it doesn't actually make sense to specify the password multiple times, however it's allowed.

Comment (vitlibar) [reply src/Parsers/ASTFunction.cpp:68]: I think I'll add more comments to the code.

File: src/Parsers/ASTFunction.cpp [hunk 4]
```diff
@@ -37,94 +37,131 @@ namespace
...
 
-        std::pair<size_t, size_t> getRange() const
+        struct Result
         {
-            if (!arguments)
-                return npos;
+            /// Result constructed by default means no arguments will be hidden.
+            size_t start = static_cast<size_t>(-1);
+            size_t count = 0; /// Mostly it's either 0 or 1. There are only a few cases where `count` can be greater than 1 (e.g. see `encrypt`).
+                              /// In all known cases secret arguments are consequent.
```
Comment (SmitaRKulkarni) [inline src/Parsers/ASTFunction.cpp:69]:
```suggestion
                              /// In all known cases secret arguments are consecutive
```

**Review by SmitaRKulkarni — APPROVED** [review]
LGTM

**Pull request merged**